File: client/android/app/src/main/java/com/stonefive/chalkak/feature/signature/SignUpViewModel.kt
package com.stonefive.chalkak.feature.signature

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.Companion.APPLICATION_KEY
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.stonefive.chalkak.ChalkakApplication
import com.stonefive.chalkak.core.ui.UiMessageEmitter
import com.stonefive.chalkak.domain.model.SocialSignUpFailure
import com.stonefive.chalkak.domain.model.SocialSignUpResult
import com.stonefive.chalkak.domain.repository.AuthRepository
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class SignUpViewModel(private val authRepository: AuthRepository) : ViewModel() {
    private val _uiState = MutableStateFlow(SignUpUiState())
    val uiState: StateFlow<SignUpUiState> = _uiState.asStateFlow()
    private val messageEmitter = UiMessageEmitter()
    val uiMessage = messageEmitter.messages

    fun completeSignUp(signaturePng: ByteArray) {
        if (_uiState.value.isSubmitting) return

        viewModelScope.launch {
            _uiState.value = SignUpUiState(status = SignUpStatus.Submitting)
            try {
                when (val result = authRepository.completeSocialSignUp(signaturePng)) {
                    is SocialSignUpResult.Success -> {
                        _uiState.value = SignUpUiState(status = SignUpStatus.Completed)
                    }

                    is SocialSignUpResult.Failure -> handleFailure(result.reason)
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Throwable) {
                showFailure("회원가입을 완료하지 못했어요. 다시 시도해 주세요.")
            }
        }
    }

    private suspend fun handleFailure(failure: SocialSignUpFailure) {
        if (
            failure == SocialSignUpFailure.REAUTHENTICATION_REQUIRED ||
            failure == SocialSignUpFailure.MISSING_LOGIN_CONTEXT
        ) {
            _uiState.value = SignUpUiState(status = SignUpStatus.ReauthenticationRequired)
            return
        }

        showFailure(failure.toMessage())
    }

    private fun showFailure(message: String) {
        _uiState.value = SignUpUiState()
        messageEmitter.showToast(message)
    }

    private fun SocialSignUpFailure.toMessage(): String = when (this) {
        SocialSignUpFailure.SIGNATURE_TOO_LARGE -> "사인 이미지가 1MB를 초과했어요."

        SocialSignUpFailure.SIGNATURE_PROCESSING_TIMEOUT -> "사인 처리에 시간이 걸리고 있어요. 다시 시도해 주세요."

        SocialSignUpFailure.SIGNATURE_NOT_FOUND -> "사인을 다시 업로드해 주세요."

        SocialSignUpFailure.INVALID_SIGNATURE -> "사용할 수 없는 사인이에요. 다시 그려주세요."

        SocialSignUpFailure.NETWORK_UNAVAILABLE -> "네트워크 연결을 확인해 주세요."

        SocialSignUpFailure.UNKNOWN -> "회원가입을 완료하지 못했어요. 다시 시도해 주세요."

        SocialSignUpFailure.MISSING_LOGIN_CONTEXT,
        SocialSignUpFailure.REAUTHENTICATION_REQUIRED,
        -> error("Handled before message mapping")
    }

    companion object {
        val Factory = viewModelFactory {
            initializer {
                val application = this[APPLICATION_KEY] as ChalkakApplication
                SignUpViewModel(application.appContainer.authRepository)
            }
        }
    }
}

File: client/iOS/chalkak/chalkak/Features/Onboarding/SignUpViewModel.swift
import Combine
import Foundation

enum SignUpStatus: Equatable {
    case idle
    case submitting
    case completed
    case reauthenticationRequired
}

struct SignUpViewState: Equatable {
    var status: SignUpStatus = .idle
    var errorMessage: String?

    var isSubmitting: Bool {
        status == .submitting
    }
}

@MainActor
final class SignUpViewModel: ObservableObject {
    @Published private(set) var state = SignUpViewState()

    private let authRepository: AuthRepository
    private let pngEncoder: any OnboardingSignaturePngEncoder
    private var signUpTask: Task<Void, Never>?

    init(
        authRepository: AuthRepository,
        pngEncoder: (any OnboardingSignaturePngEncoder)? = nil
    ) {
        self.authRepository = authRepository
        self.pngEncoder = pngEncoder ?? DefaultOnboardingSignaturePngEncoder()
    }

    deinit {
        signUpTask?.cancel()
    }

    func completeSignUp(strokes: [OnboardingSignatureStroke]) {
        guard !state.isSubmitting,
              strokes.contains(where: { !$0.isEmpty }) else { return }

        state = SignUpViewState(status: .submitting)
        signUpTask = Task { [weak self] in
            guard let self else { return }

            do {
                let signaturePNG = try pngEncoder.encode(strokes)
                let result = try await authRepository.completeSocialSignUp(
                    signaturePNG: signaturePNG
                )
                switch result {
                case .success:
                    state = SignUpViewState(status: .completed)
                case let .failure(failure):
                    handle(failure)
                }
            } catch is CancellationError {
                state = SignUpViewState()
            } catch {
                state = SignUpViewState(
                    errorMessage: "회원가입을 완료하지 못했어요. 다시 시도해 주세요."
                )
            }
        }
    }

    func dismissError() {
        state.errorMessage = nil
    }

    private func handle(_ failure: SocialSignUpFailure) {
        switch failure {
        case .reauthenticationRequired, .missingLoginContext:
            state = SignUpViewState(status: .reauthenticationRequired)
        default:
            state = SignUpViewState(errorMessage: failure.message)
        }
    }
}

private extension SocialSignUpFailure {
    var message: String {
        switch self {
        case .signatureTooLarge:
            "사인 이미지가 1MB를 초과했어요."
        case .signatureProcessingTimeout:
            "사인 처리에 시간이 걸리고 있어요. 다시 시도해 주세요."
        case .signatureNotFound:
            "사인을 다시 업로드해 주세요."
        case .invalidSignature:
            "사용할 수 없는 사인이에요. 다시 그려주세요."
        case .networkUnavailable:
            "네트워크 연결을 확인해 주세요."
        case .unknown:
            "회원가입을 완료하지 못했어요. 다시 시도해 주세요."
        case .missingLoginContext, .reauthenticationRequired:
            "로그인 정보가 만료되었어요. 다시 로그인해 주세요."
        }
    }
}
